src/bytes.rs

```diff
@@ -0,0 +1,749 @@
+            VariantMut::Spilled(spilled) => {
+                let mut ret = EcoVec::default();
+                core::mem::swap(spilled, &mut ret);
+                core::mem::forget(value);
+                ret
+            }
+        }
+    }
+}
+
+impl From<&'_ [u8]> for EcoBytes {
+    #[inline]
+    fn from(value: &'_ [u8]) -> Self {
+        Self::from_slice(value)
+    }
+}
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub(crate) struct InlineVec {
+    /// Storage!
+    buf: [u8; LIMIT],
+    /// Invariant: After masking off the tag, never exceeds LIMIT.
+    tagged_len: u8,
+}
+
+impl InlineVec {
+    #[inline]
+    pub const fn new() -> Self {
+        Self::from_buf([0; LIMIT], 0)
+    }
+
+    #[inline]
+    pub const fn from_slice(bytes: &[u8]) -> Result<Self, ()> {
+        let len = bytes.len();
+        if len > LIMIT {
+            return Err(());
+        }
+
+        let mut buf = [0; LIMIT];
+        let mut i = 0;
+        while i < len {
+            buf[i] = bytes[i];
+            i += 1;
+        }
+
+        Ok(Self::from_buf(buf, len))
+    }
+
+    #[inline]
+    pub const fn from_buf(buf: [u8; LIMIT], len: usize) -> Self {
+        Self { buf, tagged_len: len as u8 | LEN_TAG }
+    }
+
+    #[inline]
+    pub fn len(&self) -> usize {
+        usize::from(self.tagged_len & LEN_MASK)
+    }
+
+    /// The given length may not exceed LIMIT.
+    #[inline]
+    unsafe fn set_len(&mut self, len: usize) {
+        debug_assert!(len <= LIMIT);
+        self.tagged_len = len as u8 | LEN_TAG;
+    }
+
+    #[inline]
+    pub fn as_slice(&self) -> &[u8] {
+        // Safety: We have the invariant `len <= LIMIT`.
+        unsafe { self.buf.get_unchecked(..self.len()) }
+    }
+
+    #[inline]
+    pub fn as_slice_mut(&mut self) -> &mut [u8] {
+        let len = self.len();
+        // Safety: We have the invariant `len <= LIMIT`.
+        unsafe { self.buf.get_unchecked_mut(..len) }
+    }
+
+    #[inline]
+    pub fn clear(&mut self) {
+        unsafe {
+            // Safety: Trivially, `0 <= LIMIT`.
+            self.set_len(0);
+        }
+    }
+
+    #[inline]
+    pub fn push(&mut self, byte: u8) -> Result<(), ()> {
+        let len = self.len();
+        if let Some(slot) = self.buf.get_mut(len) {
+            *slot = byte;
+            unsafe {
+                // Safety: The `get_mut` call guarantees that `len < LIMIT`.
+                self.set_len(len + 1);
+            }
+            Ok(())
+        } else {
+            Err(())
+        }
+    }
+
+    #[inline]
+    pub fn pop(&mut self) -> Option<u8> {
+        let last = self.as_slice().last().copied();
+        let len = self.len().saturating_sub(1);
+        unsafe {
+            // Safety: self.len().saturating_sub(1) <= self.len()
+            self.set_len(len);
+        }
+        last
+    }
+
+    #[inline]
+    pub fn insert(&mut self, index: usize, value: u8) -> Result<(), ()> {
+        //TODO: optimize?
```

Comment:
I think a custom implementation with `ptr::copy` (like in vec.rs) does make sense here.